homeassistant/components/frank_energie/sensor.py

```diff
@@ -0,0 +1,397 @@
+    ),
+    FrankEnergieEntityDescription(
+        key="expected_costs_this_month",
+        translation_key="expected_costs_this_month",
+        device_class=SensorDeviceClass.MONETARY,
+        state_class=SensorStateClass.TOTAL,
+        native_unit_of_measurement=CURRENCY_EURO,
+        authenticated=True,
+        service_name=SERVICE_NAME_COSTS,
+        value_fn=lambda data: data.month_summary.expectedCosts,
+    ),
+    FrankEnergieEntityDescription(
+        key="invoice_previous_period",
+        translation_key="invoice_previous_period",
+        device_class=SensorDeviceClass.MONETARY,
+        state_class=SensorStateClass.TOTAL,
+        native_unit_of_measurement=CURRENCY_EURO,
+        authenticated=True,
+        service_name=SERVICE_NAME_COSTS,
+        value_fn=lambda data: data.invoices.previousPeriodInvoice.TotalAmount
+        if data.invoices.previousPeriodInvoice
+        else None,
+        attr_fn=lambda data: {
+            "Start date": data.invoices.previousPeriodInvoice.StartDate,
+            "Description": data.invoices.previousPeriodInvoice.PeriodDescription,
+        },
+    ),
+    FrankEnergieEntityDescription(
+        key="invoice_current_period",
+        translation_key="invoice_current_period",
+        device_class=SensorDeviceClass.MONETARY,
+        state_class=SensorStateClass.TOTAL,
+        native_unit_of_measurement=CURRENCY_EURO,
+        authenticated=True,
+        service_name=SERVICE_NAME_COSTS,
+        value_fn=lambda data: data.invoices.currentPeriodInvoice.TotalAmount
+        if data.invoices.currentPeriodInvoice
+        else None,
+        attr_fn=lambda data: {
+            "Start date": data.invoices.currentPeriodInvoice.StartDate,
+            "Description": data.invoices.currentPeriodInvoice.PeriodDescription,
+        },
+    ),
+    FrankEnergieEntityDescription(
+        key="invoice_upcoming_period",
+        translation_key="invoice_upcoming_period",
+        device_class=SensorDeviceClass.MONETARY,
+        state_class=SensorStateClass.TOTAL,
+        native_unit_of_measurement=CURRENCY_EURO,
+        authenticated=True,
+        service_name=SERVICE_NAME_COSTS,
+        value_fn=lambda data: data.invoices.upcomingPeriodInvoice.TotalAmount
+        if data.invoices.upcomingPeriodInvoice
+        else None,
+        attr_fn=lambda data: {
+            "Start date": data.invoices.upcomingPeriodInvoice.StartDate,
+            "Description": data.invoices.upcomingPeriodInvoice.PeriodDescription,
+        },
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Frank Energie sensor entries."""
+    frank_coordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    # Add an entity for each sensor type, when authenticated is True,
+    # only add the entity if the user is authenticated
+    async_add_entities(
+        [
```

Comment:
You can remove the square brackets, this saves a bit because it doesnt have to create a list